<h2> Комментарий ревьюера <a class="tocSkip"></h2>

Абина, привет!

Меня зовут Олег Михеев, и я буду проверять твой проект. Предлагаю общаться на "ты", но если тебе будет некомфортно, обязательно сообщи, и мы перейдем на "вы". Моя главная цель - не указать на твои ошибки, а поделиться опытом и подготовить тебя к работе аналитиком. Увидев у тебя ошибку, я постараюсь указать на её наличие и дам возможность найти и самостоятельно исправить её.

В проверке проекта я буду использовать разные цвета для обозначения комментариев. Например:

<br/>

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Так выделены удачные решения, на которые можно опираться в будущем.
</div>

<br/>

<div class="alert alert-warning">
    <h3> Комментарий ревьюера ⚠️<a class="tocSkip"> </h3>

Когда решение на отдельном шаге станет лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h3> Комментарий ревьюера ❌<a class="tocSkip"></h3>

 А так выделены решения, которые необходимо доработать, чтобы я мог принять проект.
</div>
    
  
Если вносишь изменения в проект, указывай, пожалуйста, это в своих комментариях. Будет удобно, если обозначишь их заметным цветом. Например, вот так:
       
<div class="alert alert-block alert-info">
    
<h3> Комментарий студента  <a class="tocSkip"></h3>
    
Можешь кликнуть сюда два раза и скопировать код комментария. Также если что-то непонятно или остались вопросы по ревью, обязательно пиши.
    
</div>
    
Пожалуйста, не удаляй и не перемещай мои комментарии, чтобы не произошло путаницы, и я мог быстрее проверить проект.
<br/>

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по ревью**

Здорово, что удалось пройти по всем этапам исследования. Теория усвоена уже на хорошем уровне, но некоторые моменты необходимо доработать, чтобы можно было идти дальше.

**Что нужно сделать, чтобы я мог принять проект**:

* Вывести сводные таблицы с количеством клиентов, количеством должников и процентом должников для всех вопросов исследовательской части.
* Пересмотреть выводы там, где необходимо
* Написать подробный финальный вывод, учитывающий все группы клиентов с процентными показателями и информацию о предобработке

**Успехов! Жду проект на повторное ревью!**

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по ревью v2**

Cпасибо, что учла все замечания.

Очень качественно выполненный проект! Мне было приятно с тобой работать 🤝

Видно, что ты последовательна и вдумчива в своих выводах.
    
А проект написан аккуратно, поэтому его легко воспринимать.

Продолжай в том же духе и станешь крутым специалистом.

**Успехов в дальнейшем обучении!**

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [2]:
import pandas as pd

data = pd.read_csv('data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [3]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

## Предобработка данных

### Удаление пропусков

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [8]:
for income_type in data['income_type'].unique():
    median_income = data.loc[data['income_type'] == income_type, 'total_income'].median()
    data.loc[(data['income_type'] == income_type) & (data['total_income'].isna()), 'total_income'] = median_income

In [9]:
# Проверка
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [10]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [11]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [12]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [13]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [14]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [15]:
for income_type in data['income_type'].unique():
    median_days_employed = data.loc[data['income_type'] == income_type, 'days_employed'].median()
    data.loc[(data['income_type'] == income_type) & (data['days_employed'].isna()), 'days_employed'] = median_days_employed

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [16]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [18]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [19]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [20]:
data.duplicated().sum()

71

In [21]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [22]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [23]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [24]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [25]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [26]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

 <div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

Чтобы не приходилось несколько раз выполнять одну и ту же операцию, можно автоматизировать процесс вывода сводных таблиц с помощью пользовательской функции. Например, такой:

 ```py
def p_table_create(category):
  table = data.groupby(category)['debt'].agg(['count', 'sum', 'mean'])\
    .sort_values(by = 'mean', ascending = False)  
  table['mean'] = table['mean'].apply(lambda x: "{0:.2%}".format(x))
  return table
```
    
</div>

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [27]:
children_pivot_table = (data.groupby('children')
                 .agg(total_clients=('debt', 'count'),             
                     total_debtors=('debt', 'sum'))               
                 .assign(prctg_debtors=lambda x: x['total_debtors'] / x['total_clients'] * 100)  
              )

print(children_pivot_table)

          total_clients  total_debtors  prctg_debtors
children                                             
0                 14091           1063       7.543822
1                  4808            444       9.234609
2                  2052            194       9.454191
3                   330             27       8.181818
4                    41              4       9.756098
5                     9              0       0.000000


**Вывод:** 

Семьи с четырями, двумя детьми и семьи с одним ребенком имеют самый высокий процент должников (9,75%, 9,45% и 9,23% соответственно). Семьи без детей имеют самый низкий процент должников (7,54%). Где-то посередине находятся семьи с тремя детьми (8,18%). Да, есть зависимость между количеством детей и возвратом кредита в срок. У многодетных семей процент должников выше, чем у семей без детей. 

Мало данных про семей с пятью, четыремя и тремя детьми. Тогда можно сравнить, другие три категории: у семей с одним или двумя детьми процент задолжников больше, чем у семей без детей. Значит, что семьи с детьми имеют трудности с погашением кредита и чаще всего не оплачивают в срок.

 <div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

 Вывод абсолютно верный!
</div>

<div class="alert alert-danger">
    <h3> Комментарий ревьюера ❌<a class="tocSkip"> </h3>

Ты уже понимаешь, как можно группировать данные и умело пользуешься функциями.        
        
Однако с помощью count() удалось посчитать количество клиентов по группам. Эта информация пригодится, но по ней нельзя сделать вывод о зависимости. В данном же случае необходимо сравнить процент должников по группам, учитывая количество клиентов в каждой группе, чтобы убедиться в объективности выводов.
        
В теории вы проходили сводные таблицы, и я хочу убедиться, что тебе удалось их освоить на хорошем уровне. Потому что в дальнейшем ты будешь пользоваться ими постоянно. Выведи, пожалуйста, общую сводную таблицу с количеством клиентов, количеством должников и процентом должников, используя несколько функций одновременно.

Даю несколько ссылок, в которых есть примеры, как это можно сделать:       
* С помощью pivot_table и аргумента aggfunc https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html
* С помощью groupby и функции .agg() https://stackoverflow.com/questions/12589481/multiple-aggregations-of-the-same-column-using-pandas-groupby-agg

Добавь такие сводные таблицы для всех вопросов исследовательской части и напиши вывод по новой информации, пожалуйста.
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [32]:
family_status_pivot_table = (data.groupby('family_status')
                 .agg(total_clients=('debt', 'count'),             
                     total_debtors=('debt', 'sum'))               
                 .assign(prctg_debtors=lambda x: x['total_debtors'] / x['total_clients'] * 100)  
              )

print(family_status_pivot_table)

                       total_clients  total_debtors  prctg_debtors
family_status                                                     
Не женат / не замужем           2796            273       9.763948
в разводе                       1189             84       7.064760
вдовец / вдова                   951             63       6.624606
гражданский брак                4134            385       9.313014
женат / замужем                12261            927       7.560558


**Вывод:** 

Самый высокий процент должников среди людей с семейным положением `не женат/не замужем` и `гражданский брак` - 9,76% и 9,31% соответственно. `Вдовец/вдова` имеют самый низкий процент должников (6.62%). `В разводе` и `женат/замужем` находятся где-то между ними (7,06% и 7,56%). Клиенты банка с семейным статусом "не женат/ не замужем" и "гражданский брак" чаще всего становятся должниками. 

 <div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

 Верно. А по какому признаку можно разделить все выборки? Какой фактор может влиять на возвратность кредитов, как ты считаешь?

 Если захочешь округлить значения столбца до нужного количества знаков после запятой, можно воспользоваться функций .round().
    
Например, такой код округлит столбец до трёх знаков после запятой:
    
    data['column'] = data['column'].round(3)
</div>

<div class="alert alert-block alert-danger">
<h3> Комментарий ревьюера ❌<a class="tocSkip"></h3>

Аналогично предыдущему шагу. По данной информации сделать вывод о зависимости нельзя.
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [34]:
income_pivot_table = (data.groupby('total_income_category')
                 .agg(total_clients=('debt', 'count'),             
                     total_debtors=('debt', 'sum'))               
                 .assign(prctg_debtors=lambda x: x['total_debtors'] / x['total_clients'] * 100)  
              )

print(income_pivot_table)

                       total_clients  total_debtors  prctg_debtors
total_income_category                                             
A                                 25              2       8.000000
B                               5014            354       7.060231
C                              15921           1353       8.498210
D                                349             21       6.017192
E                                 22              2       9.090909


**Вывод:** 

Напомним, что столбец `total_income_category` был составлен согласно следующим диапазонам:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

Больше всего должников среди людей с доходом категории `E` (самый маленький доход в выборке) - 9,09%. Категория `C` имеет второй самый большой процент по количеству должников (8,49 %). Люди с доходом категории `D` имеют самый низкий процент (6,01%). Где-то между ними находятся категории `B` и `A`, люди с самыми большими доходами. Однако, количество клиентов в категории A очень мало (25), поэтому этот показатель может быть нерепрезентативным. То же самое и с категорией `E` где, всего два задолжника, поэтому нельзя сказать, что у людей с самым маленьким доходом чаще всего проблемы с возрастом кредита. Можно предположить, что зависимость между уровнем дохода и возвратом кредита в срок существует. Опираясь на данные доходов категорий B, C, D сделаем вывод, что должниками скорее становятся клиенты категории C, а более отвественно подходят к делу клиенты категории D. 


<div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

Верно. Правильно, что обращаешь внимание на объёмы категорий клиентов. Так можно быть уверенным в объективности выводов.
    
Факультативно, можно попробовать решить задание альтернативным способом, разделив выборку на равные по объёму части, в данном случае, по столбцу сумм дохода. Для этого можно воспользоваться функцией pd.qcut().
    
Здесь можно об этом почитать:
    
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html

Метод qcut() делит числовой столбец на равные по объёму части на основании квантилей. Квантили — это статистические показатели, которые разделяют набор данных на равные части. Квантили включают квартили (25%, 50% и 75%), децили (каждые 10%) и процентили (каждые 1%). Приведу пример разделения на 3 равных отрезка, находящихся на отметках 33,3% и 66,7%.
        
    pd.qcut(data['column'], q=3, labels=['one', 'two', 'three'])
        
Можно не указывать названия отрезков с помощью labels. Тогда названия отрезкам присвоятся в численном виде.
</div>

<div class="alert alert-block alert-danger">
<h3> Комментарий ревьюера ❌<a class="tocSkip"></h3>

Аналогично предыдущим шагам. И на этапе предобработки был создан столбец с категориями дохода - total_income_category, по которому и нужно сделать анализ на этом шаге. Доработай эти моменты, пожалуйста.
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [36]:
purpose_pivot_table = (data.groupby('purpose_category')
                 .agg(total_clients=('debt', 'count'),             
                     total_debtors=('debt', 'sum'))               
                 .assign(prctg_debtors=lambda x: x['total_debtors'] / x['total_clients'] * 100)  
              )

print(purpose_pivot_table)

                          total_clients  total_debtors  prctg_debtors
purpose_category                                                     
операции с автомобилем             4279            400       9.347978
операции с недвижимостью          10751            780       7.255139
получение образования              3988            369       9.252758
проведение свадьбы                 2313            183       7.911803


**Вывод:**

"Операции с автомобилем" и "Получение образования" имеют самый высокий процент должников (9.35% и 9.25% соответственно). "Операции с недвижимостью" имеет наименьший процент должников (7.26%). "Проведение свадьбы" находится посередине (7.91%). Можно предположить, что есть некая зависимость между целью кредита и его возвратом в срок. Кредиты, взятые на получение образования и операции с автомобилем, чаще всего не оплачиваются в срок. 

<div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

С выводом согласен. Попробуй предположить, почему операции на недвижимость и свадьбы выплачиваются.
</div>

<div class="alert alert-block alert-danger">
<h3> Комментарий ревьюера ❌<a class="tocSkip"></h3>

Аналогично предыдущим шагам
</div>

**Вывод:** 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

In [26]:
data.isna().sum()

children               0
days_employed       2162
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2162
purpose                0
purpose_category       0
dtype: int64

In [28]:
data.groupby('income_type')['total_income'].mean().sort_values()

income_type
в декрете           53829.130729
студент             98201.625314
безработный        131339.751676
пенсионер          137198.468121
сотрудник          161365.012384
госслужащий        170743.143230
компаньон          202585.823720
предприниматель    499163.144947
Name: total_income, dtype: float64

*Ответ:* 

1. Ошибки при вводе данных могут возникнуть как по невнимательности
2. В выборке есть женщины в декрете, студенты и безработных. В среднем у таких людей доход меньше других. Это может быть связано с тем, что у некоторых людей из этих групп нет работы, значит нет дохода. По этой же и причине, в столбце `days_employed` есть пропущенные значения. 
3. Ошибки при импортировании данных. При импорте данных из разных источников могут возникнуть ошибки, из-за чего возникают пропуски. Другие технические проблемы
4. Неправильный ввод данных 

<div class="alert alert-warning">
    <h3> Комментарий ревьюера ⚠️<a class="tocSkip"> </h3>

Верные примеры, намеренное или ненамеренное сокрытие информации - это человеческий фактор. А какую ещё причину пропусков, не связанную напрямую с человеком, ты можешь назвать?
</div>

 <div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

 Верно, второй причиной пропусков являются проблемы технического характера.
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Медиана  — это центральное значение в отсортированном наборе данных. Она не чувствительна к выбросам, как среднее значение, поэтому этот метод удобнее использовать, чтобы справиться с пропущенными значениями. Заполнение медианой помогает сохранить исходное распределение данных, так как  новые значения не смещают его в сторону.

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Верно, медиана была выбрана именно по причине устойчивости к выбросам
</div>

### Шаг 4: общий вывод.

**Предобработка данных** 

В столбцах `days_employed` и `total_income` были пропущенные значения, которые мы заменили медианными значениями по каждому типу занятости. В столбце `children` были аномальные значения, которые мы удалили. Удалили неявные дубликаты в столбце `education` и дубликаты во всех данных. Категоризировали данные в столбцах `purpose` и `total_income`. 

**Влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок?**

1. Количество детей в семье влияет на погашение кредита в срок. Семьи с детьми более подвержены риску стать должниками, когда как клиенты банка без детей имеют наименьший процент задолжников. 
2. Не женатые/не замужние клиенты и пары, состоящие в гражданском браке, более тяготеют к риску стать задолжниками 
3. Клиенты банка с доходом категории E (50001–200000) склонны к тому, чтобы стать должниками. 
4. Кредиты на операции с автомобилем и получение образования чаще всего не оплачиваются в срок. 

<div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

Теперь получился развёрнутый хороший вывод, учитывающий все показатели и этапы проекта. Отличная работа!
    
</div>

<div class="alert alert-danger">
    <h3> Комментарий ревьюера ❌<a class="tocSkip"> </h3>
    
Все этапы исследовательской части рассмотрены. Но не хватает деталей.
        
Стоит учитывать, что твой потенциальный руководитель может обратить внимание только на общий вывод. Поэтому важно отразить в нём все основные результаты из всех шагов проекта в числах. А также описать проблемы, найденные на этапе предобработки, и пути их решения
        
Можно опираться на такую структуру финального вывода во всех проектах:
* предобработка данных с описанием найденных проблем и путей их решения
* ответы на цели проекта
* рекомендации для заказчика

Чтобы лучше понимать, как оформлять комментарии в markdown-ячейках, советую изучить [краткое руководство по Маркдауну](https://paulradzkov.com/2014/markdown_cheatsheet/).

</div>